desc/basis.py

```diff
@@ -1413,83 +1412,150 @@ def zernike_radial_poly(r, l, m, dr=0, exact="auto"):
+        P_n1 = P_past[1]  # Jacobi at N-1
+        P_n = jacobi_poly_single(r_jacobi, N, alpha, 0, P_n1, P_n2)
+
+        # Calculate Radial part of Zernike for N,alpha
+        result = (-1) ** N * r**alpha * P_n
+        # Check if the calculated values is in the given modes
+        _, _, _, _, _, out = fori_loop(
+            0, m.size, update_zernike_output, (m, n, alpha, N, result, out)
         )
-    else:
-        raise NotImplementedError(
-            "Analytic radial derivatives of Zernike polynomials for order>4 "
-            + "have not been implemented."
+
+        # Shift past values if needed
+        # For derivative order dx, if N is smaller than 2+dx, then only the initial
+        # value calculated by find_init_jacobi function will be used. So, if you update
+        # P_n's, preceeding values will be wrong.
+        mask = N >= 2
+        P_n2 = jnp.where(mask, P_n1, P_n2)
+        P_n1 = jnp.where(mask, P_n, P_n1)
+        # Form updated P_past matrix
+        P_past = P_past.at[0].set(P_n2)
+        P_past = P_past.at[1].set(P_n1)
+
+        return (alpha, out, P_past)
+
+    def body(alpha, out):
+        # find l values with m values equal to alpha
+        l_alpha = jnp.where(m == alpha, l, 0)
+        # find the maximum among them
+        L_max = jnp.max(l_alpha)
+        # Maximum possible value for n for loop bound
+        N_max = (L_max - alpha) // 2
+
+        # First 2 Jacobi Polynomials (they don't need recursion)
+        # P_past stores last 2 Jacobi polynomials (and required derivatives)
+        # evaluated at given r points
+        P_past = jnp.zeros(2)
+        P_past = P_past.at[0].set(jacobi_poly_single(r_jacobi, 0, alpha, beta=0))
+        # Jacobi for n=1
+        P_past = P_past.at[1].set(jacobi_poly_single(r_jacobi, 1, alpha, beta=0))
+
+        # Loop over every n value
+        _, out, _ = fori_loop(
+            0, (N_max + 1).astype(int), body_inner, (alpha, out, P_past)
         )
-    return s * jnp.where((l - m) % 2 == 0, out, 0)
+        return out
+
+    # Make inputs 1D arrays in case they aren't
+    m = jnp.atleast_1d(m)
+    l = jnp.atleast_1d(l)
+
+    # From the vectorization, the overall output will be (r.size, m.size)
+    out = jnp.zeros(m.size)
+    r_jacobi = 1 - 2 * r**2
+    m = jnp.abs(m)
+    n = ((l - m) // 2).astype(int)
+
+    M_max = jnp.max(m)
+    # Loop over every different m value. There is another nested
+    # loop which will execute necessary n values.
+    out = fori_loop(0, (M_max + 1).astype(int), body, (out))
+    return out
+
+
+def jacobi_poly_single(x, n, alpha, beta=0, P_n1=0, P_n2=0):
+    """Evaluate Jacobi for single alpha and n pair."""
+    c = 2 * n + alpha + beta
+    a1 = 2 * n * (c - n) * (c - 2)
+    a2 = (c - 1) * (c * (c - 2) * x + (alpha - beta) * (alpha + beta))
+    a3 = 2 * (n + alpha - 1) * (n + beta - 1) * c
+
+    # Check if a1 is 0, to prevent division by 0
+    a1 = jnp.where(a1 == 0, 1e-6, a1)
```

Comment:
can `safediv` be used instead of this?

Reply:
Hmm, safediv has extra checks, I don't think we need that everytime this function is called. jnp.where for single argument is fast but still takes time, especially if the intended value is in the last argument, (for larger array conditions, I realized if 3rd instead of 2nd entry is used most of the time, it takes significantly more time). I am not sure about how much difference would it make for this case, but still this line is just for not throwing error in vain, only possible a1==0 cases are n=0 or 1, and they are replaced right at the end, it is dummy calculation anyway

Reply:
```python
c = 2 * n + alpha + beta
a1 = 2 * n * (c - n) * (c - 2)
```
a1=0 for n=0, c=n or c=2,
c=n :
    n = -(alpha+beta) where alpha and beta both positive so impossible

n=2:
   n=0 and alpha+beta=2 or n=1 and alpha=beta=0

Reply:
so are you saying this line is not needed? But anyhow If this is faster than safediv then ok that is fine, we can keep it as is, thanks for the explanaiton

Reply:
yeah, it is kind of not needed, some tests were failing, so I added it.